codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/EndpointGenerator.java

```diff
@@ -0,0 +1,456 @@
+        });
+        w.write("");
+        w.write("var endpoint $T", SymbolUtils.createValueSymbolBuilder("Endpoint", AwsGoDependency.AWS_CORE)
+                .build());
+        w.write("endpoint, err = m.Resolver.ResolveEndpoint(awsmiddleware.GetRegion(ctx))");
+        w.openBlock("if err != nil {", "}", () -> {
+            w.write("return out, metadata, fmt.Errorf(\"failed to resolve service endpoint\")");
+        });
+        w.write("");
+        w.write("req.URL, err = url.Parse(endpoint.URL)");
+        w.openBlock("if err != nil {", "}", () -> {
+            w.write("return out, metadata, fmt.Errorf(\"failed to parse endpoint URL: %w\", err)");
+        });
+        w.write("");
+        w.openBlock("if len(awsmiddleware.GetSigningName(ctx)) == 0 {", "}", () -> {
+            w.write("signingName := endpoint.SigningName");
+            w.openBlock("if len(signingName) == 0 {", "}", () -> {
+                w.write("signingName = $S", serviceShape.expectTrait(ServiceTrait.class).getArnNamespace());
+            });
+            w.write("ctx = awsmiddleware.SetSigningName(ctx, signingName)");
+        });
+        w.write("ctx = awsmiddleware.SetSigningRegion(ctx, endpoint.SigningRegion)");
+        w.write("");
+        w.write("return next.HandleSerialize(ctx, in)");
+    }
+
+    private void generateMiddlewareStructureMembers(GoStackStepMiddlewareGenerator g, GoWriter w) {
+        w.write("Resolver $L", getResolverInterfaceName());
+    }
+
+    private String getPublicResolverConstructorName() {
+        return "NewDefaultEndpointResolver";
+    }
+
+    private Symbol.Builder getInternalEndpointsSymbol(String symbolName, boolean pointable) {
+        Symbol.Builder builder;
+        if (pointable) {
+            builder = SymbolUtils.createPointableSymbolBuilder(symbolName);
+        } else {
+            builder = SymbolUtils.createValueSymbolBuilder(symbolName);
+        }
+        return builder.namespace(settings.getModuleName() + "/" + getInternalEndpointsPath(), "/");
+    }
+
+    private Symbol.Builder getInternalEndpointsSymbol(String symbolName, boolean pointable, String moduleAlias) {
+        return getInternalEndpointsSymbol(symbolName, pointable)
+                .putProperty(SymbolUtils.NAMESPACE_ALIAS, moduleAlias);
+    }
+
+    private void generatePublicResolverTypes(GoWriter writer) {
+        final String internalAlias = "serviceEndpoints";
+
+        Symbol awsEndpointSymbol = SymbolUtils.createValueSymbolBuilder("Endpoint", AwsGoDependency.AWS_CORE).build();
+        Symbol internalEndpointsSymbol = getInternalEndpointsSymbol(getResolverImplementationName(), true,
+                internalAlias).build();
+
+        // Generate Resolver Interface
+        writer.openBlock("type $L interface {", "}", getResolverInterfaceName(), () -> {
+            writer.write("ResolveEndpoint(region string) ($T, error)", awsEndpointSymbol);
+        });
```

Comment:
For future capabilities does it make sense for the ResolveEndpoint to take the same options as the EndpointResolver interface? Also allows the same resolver value to be shared across all operations